From ee4176898223eae4ab1fcf31dc222e731ff68444 Mon Sep 17 00:00:00 2001
From: "Edward.Lee im105" <edward.lee@augentix.com>
Date: Wed, 21 Jun 2023 16:29:36 +0800
Subject: [PATCH 6/8] Filter out ops not used by ShuffleNetV2 (Associates:
 #47451)

For now (size-optim) TFLite is tailored for ShufflenetV2. Other models
should link full TFLite by default. To reduce storage size, we
filter out unused ops from size optimized TFLite.
---
 Makefile.lite                       | 18 +++++++++---------
 tensorflow/lite/kernels/register.cc | 22 +++++++++-------------
 2 files changed, 18 insertions(+), 22 deletions(-)

diff --git a/Makefile.lite b/Makefile.lite
index b982ed72..d2f03b83 100644
--- a/Makefile.lite
+++ b/Makefile.lite
@@ -192,8 +192,11 @@ $(root)/tensorflow/lite/kernels/unique.cc \
 $(root)/tensorflow/lite/kernels/pack.cc \
 $(root)/tensorflow/lite/kernels/unpack.cc \
 $(root)/tensorflow/lite/kernels/zeros_like.cc \
+$(root)/tensorflow/lite/kernels/sub.cc \
+$(root)/tensorflow/lite/kernels/add.cc \
 $(root)/tensorflow/lite/kernels/sparse_to_dense.cc \
 $(root)/tensorflow/lite/experimental/resource/static_hashtable.cc \
+$(root)/tensorflow/lite/kernels/resize_bilinear.cc \
 $(root)/tensorflow/lite/kernels/pooling3d.cc \
 $(root)/tensorflow/lite/kernels/scatter_nd.cc \
 $(root)/tensorflow/lite/kernels/space_to_batch_nd.cc \
@@ -208,21 +211,18 @@ $(root)/tensorflow/lite/kernels/atan2.cc \
 $(root)/tensorflow/lite/kernels/sign.cc \
 $(root)/tensorflow/lite/kernels/exp.cc \
 $(root)/tensorflow/lite/kernels/shape.cc \
+$(root)/tensorflow/lite/kernels/neg.cc \
 $(root)/tensorflow/lite/kernels/squeeze.cc \
 $(root)/tensorflow/lite/tools/optimize/sparsity/format_converter.cc \
+$(root)/tensorflow/lite/kernels/mul.cc \
 $(root)/tensorflow/lite/kernels/maximum_minimum.cc \
+$(root)/tensorflow/lite/kernels/div.cc \
 $(root)/tensorflow/lite/kernels/batch_matmul.cc \
 $(shell find tensorflow/lite/tools/make/downloads/absl/absl/ \
-	             -type f -name \*.cc)
+	             -type f -name \*.cc) \
+$(root)/tensorflow/lite/kernels/elementwise.cc \
 
 CORE_CC_KERNEL_FILTER_SRCS_UNVERIFIED += \
-$(root)/tensorflow/lite/kernels/elementwise.cc \
-$(root)/tensorflow/lite/kernels/resize_bilinear.cc \
-$(root)/tensorflow/lite/kernels/neg.cc \
-$(root)/tensorflow/lite/kernels/sub.cc \
-$(root)/tensorflow/lite/kernels/mul.cc \
-$(root)/tensorflow/lite/kernels/add.cc \
-$(root)/tensorflow/lite/kernels/div.cc \
 $(root)/tensorflow/lite/optional_debug_tools.cc \
 $(root)/tensorflow/lite/profiling/memory_info.cc \
 $(root)/tensorflow/lite/profiling/platform_profiler.cc \
@@ -600,7 +600,7 @@ call:
 	@echo ========tf lib obj:: $(LIB_OBJS)
 
 mask:
-	@echo ======== $(TARGET_LITE_INC) 
+	@echo ======== $(TARGET_LITE_INC)
 	@echo ======== $(TARGET_ABSL_INC)
 	@echo ======== $(FLATBUFFERS_INCLUDES)
 	@echo ======== $(MAKEFILE_DIR)/downloads/flatbuffers/include/ $($INCDIR)/
diff --git a/tensorflow/lite/kernels/register.cc b/tensorflow/lite/kernels/register.cc
index 027dc589..5f76e2c0 100644
--- a/tensorflow/lite/kernels/register.cc
+++ b/tensorflow/lite/kernels/register.cc
@@ -34,9 +34,9 @@ TfLiteRegistration* Register_DETECTION_POSTPROCESS();
 namespace builtin {
 
 BuiltinOpResolver::BuiltinOpResolver() {
+#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_ABS, Register_ABS(), /* min_version = */ 1,
             /* max_version = */ 4);
-#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_HARD_SWISH, Register_HARD_SWISH());
 #endif // TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_RELU, Register_RELU(), /* min_version = */ 1,
@@ -47,10 +47,10 @@ BuiltinOpResolver::BuiltinOpResolver() {
             /* max_version = */ 3);
   AddBuiltin(BuiltinOperator_TANH, Register_TANH(), /* min_version = */ 1,
             /* max_version = */ 3);
-#endif // TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_LOGISTIC, Register_LOGISTIC(),
             /* min_version = */ 1,
             /* max_version = */ 3);
+#endif // TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_AVERAGE_POOL_2D, Register_AVERAGE_POOL_2D(),
              /* min_version */ 1,
              /* max_version */ 3);
@@ -93,27 +93,25 @@ BuiltinOpResolver::BuiltinOpResolver() {
 #ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_LSH_PROJECTION, Register_LSH_PROJECTION());
   AddBuiltin(BuiltinOperator_HASHTABLE_LOOKUP, Register_HASHTABLE_LOOKUP());
-#endif // TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_SOFTMAX, Register_SOFTMAX(),
              /* min_version = */ 1,
              /* max_version = */ 3);
+#endif // TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_CONCATENATION, Register_CONCATENATION(),
              /* min_version = */ 1,
              /* max_version = */ 3);
+#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_ADD, Register_ADD(),
             /* min_version */ 1,
             /* max_version */ 4);
-#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_SPACE_TO_BATCH_ND, Register_SPACE_TO_BATCH_ND(),
             /* min_version = */ 1,
             /* max_version = */ 3);
   AddBuiltin(BuiltinOperator_BATCH_TO_SPACE_ND, Register_BATCH_TO_SPACE_ND(),
             /* min_version = */ 1,
             /* max_version = */ 3);
-#endif // TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_MUL, Register_MUL(), /* min_version = */ 1,
              /* max_version = */ 5);
-#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_L2_NORMALIZATION, Register_L2_NORMALIZATION(),
             /* min_version = */ 1,
             /* max_version = */ 2);
@@ -130,13 +128,15 @@ BuiltinOpResolver::BuiltinOpResolver() {
 #endif // TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_PAD, Register_PAD(), /* min_version = */ 1,
              /* max_version = */ 4);
+#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_PADV2, Register_PADV2(), /* min_version = */ 1,
              /* max_version = */ 4);
+#endif // TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_RESHAPE, Register_RESHAPE());
+#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_RESIZE_BILINEAR, Register_RESIZE_BILINEAR(),
             /* min_version = */ 1,
             /* max_version = */ 4);
-#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_RESIZE_NEAREST_NEIGHBOR,
             Register_RESIZE_NEAREST_NEIGHBOR(),
             /* min_version = */ 1,
@@ -158,6 +158,7 @@ BuiltinOpResolver::BuiltinOpResolver() {
   AddBuiltin(BuiltinOperator_MEAN, Register_MEAN(),
              /* min_version = */ 1,
              /* max_version = */ 3);
+#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_DIV, Register_DIV(),
             /* min_version */ 1,
             /* max_version */ 2);
@@ -170,7 +171,6 @@ BuiltinOpResolver::BuiltinOpResolver() {
   AddBuiltin(BuiltinOperator_SPLIT_V, Register_SPLIT_V(),
              /* min_version = */ 1,
              /* max_version = */ 2);
-#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_SQUEEZE, Register_SQUEEZE(),
             /* min_version = */ 1,
             /* max_version = */ 2);
@@ -192,8 +192,8 @@ BuiltinOpResolver::BuiltinOpResolver() {
   AddBuiltin(BuiltinOperator_DEQUANTIZE, Register_DEQUANTIZE(),
              /* min_version = */ 1,
              /* max_version = */ 4);
-  AddBuiltin(BuiltinOperator_PRELU, Register_PRELU());
 #ifndef TFLITE_OPTIM
+  AddBuiltin(BuiltinOperator_PRELU, Register_PRELU());
   AddBuiltin(BuiltinOperator_MAXIMUM, Register_MAXIMUM(),
             /* min_version = */ 1,
             /* max_version = */ 4);
@@ -221,9 +221,7 @@ BuiltinOpResolver::BuiltinOpResolver() {
   AddBuiltin(BuiltinOperator_FLOOR, Register_FLOOR());
   AddBuiltin(BuiltinOperator_CEIL, Register_CEIL());
   AddBuiltin(BuiltinOperator_ROUND, Register_ROUND());
-#endif // TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_NEG, Register_NEG());
-#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_SELECT, Register_SELECT(),
             /* min_version = */ 1,
             /* max_version = */ 2);
@@ -233,11 +231,9 @@ BuiltinOpResolver::BuiltinOpResolver() {
             /* max_version = */ 5);
   AddBuiltin(BuiltinOperator_SIN, Register_SIN());
   AddBuiltin(BuiltinOperator_COS, Register_COS());
-#endif // TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_TRANSPOSE_CONV, Register_TRANSPOSE_CONV(),
              /* min_version = */ 1,
              /* max_version = */ 3);
-#ifndef TFLITE_OPTIM
   AddBuiltin(BuiltinOperator_TILE, Register_TILE(),
             /* min_version = */ 1,
             /* max_version = */ 2);
-- 
2.47.1

